govcd/vapp_network.go

```diff
@@ -0,0 +1,137 @@
+	util.Logger.Printf("[TRACE] UpdateNetworkFirewallRulesAsync with values: id: %s and firewallServiceConfiguration: %#v", networkId, firewallRules)
+	uuid := extractUuid(networkId)
+	networkToUpdate, err := vapp.GetVappNetworkById(uuid, true)
+	if err != nil {
+		return Task{}, err
+	}
+
+	if networkToUpdate.Configuration.Features == nil {
+		networkToUpdate.Configuration.Features = &types.NetworkFeatures{}
+	}
+	networkToUpdate.Xmlns = types.XMLNamespaceVCloud
+
+	if networkToUpdate.Configuration.Features.FirewallService == nil {
+		return Task{}, fmt.Errorf("provided network isn't connecd to org network or isn't fenced")
+	}
+	networkToUpdate.Configuration.Features.FirewallService.LogDefaultAction = logDefaultAction
+	networkToUpdate.Configuration.Features.FirewallService.DefaultAction = defaultAction
+	networkToUpdate.Configuration.Features.FirewallService.FirewallRule = firewallRules
+
+	// here we use `PUT /network/{id}` which allow to change vApp network.
+	// But `GET /network/{id}` can return org VDC network or vApp network.
+	apiEndpoint := vapp.client.VCDHREF
+	apiEndpoint.Path += "/network/" + uuid
+
+	return vapp.client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodPut,
+		types.MimeVappNetwork, "error updating vApp Network firewall rules: %s", networkToUpdate)
+}
+
+// GetVappNetworkById returns a VApp network reference if the vApp network ID matches an existing one.
+// If no valid VApp network is found, it returns a nil VApp network reference and an error
+func (vapp *VApp) GetVappNetworkById(id string, refresh bool) (*types.VAppNetwork, error) {
+	util.Logger.Printf("[TRACE] [GetVappNetworkById] getting vApp Network: %s and refresh %t", id, refresh)
+
+	if refresh {
+		err := vapp.Refresh()
+		if err != nil {
+			return nil, fmt.Errorf("error refreshing vapp: %s", err)
+		}
+	}
+
+	//vApp Might Not Have Any networks
+	if vapp.VApp.NetworkConfigSection == nil || len(vapp.VApp.NetworkConfigSection.NetworkConfig) == 0 {
+		return nil, ErrorEntityNotFound
+	}
+
+	util.Logger.Printf("[TRACE] Looking for networks: %s --- %d", id, len(vapp.VApp.NetworkConfigSection.NetworkConfig))
+	for _, vappNetwork := range vapp.VApp.NetworkConfigSection.NetworkConfig {
+		util.Logger.Printf("[TRACE] Looking at: %s", vappNetwork.Link.HREF)
+		if equalIds(id, vappNetwork.ID, vappNetwork.Link.HREF) {
+			vappNetwork := &types.VAppNetwork{}
+
+			apiEndpoint := vapp.client.VCDHREF
+			apiEndpoint.Path += "/network/" + extractUuid(id)
+
+			_, err := vapp.client.ExecuteRequest(apiEndpoint.String(), http.MethodGet,
+				types.MimeVappNetwork, "error getting vApp network: %s", nil, vappNetwork)
+			if err != nil {
+				return nil, err
+			}
+			return vappNetwork, nil
+		}
+	}
+	util.Logger.Printf("[TRACE] GetVappNetworkById returns not found entity")
+	return nil, ErrorEntityNotFound
+}
+
+// GetVMByName returns a VM reference if the VM name matches an existing one.
```

Comment:
Wrong function and Entity names in the comment

Reply:
updated